include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   * quadrature points; (ii) project the data to the FiniteElement space
+   * and thereby store results in global vectors; (iii) use parallel::distributed::SolutionTransfer
+   * to project FE vectors to the new mesh; and (iv) finally project the
+   * data back to the quadrature points on the new mesh via FEValues class.
+   * The ContinuousQuadratureDataTransfer class aims at simplifying the whole
+   * process by only requiring that the quadrature point data class is derived
+   * from the TransferableQuadraturePointData. Everything else will be done automatically.
+   *
+   * @note This class is not well suited for the transfer of discontinuous
+   * quadrature point data. For example for the elasto-plastic constitutive model
+   * on a cell only some quadrature points may be in the plastic state whereas
+   * others remain elastic. In this scenario using this class to transfer
+   * plastic strain at quadrature points does not make much sense as this field
+   * is discontinuous, whereas cell-based L2-projection performed by this class
+   * will smooth this discontinuity across all child elements.
+   *
+   * @author Denis Davydov, Jean-Paul Pelteret, 2016
+   */
+  template <int dim, typename DataType>
+  class ContinuousQuadratureDataTransfer
+  {
+  public:
+    static_assert(std::is_base_of<TransferableQuadraturePointData, DataType>::value,
+                  "User's DataType class should be derived from QPData");
+
+    /**
+     * A typedef for a cell.
+     */
+    typedef typename parallel::distributed::Triangulation<dim>::cell_iterator CellIteratorType;
+
+    /**
+     * Constructor which takes the FiniteElement @p projection_fe , the quadrature
+     * rule @p mass_quadrature used to integrate its local mass matrix and
+     * finally the quadrature rule @p data_quadrature which is used to store @p DataType .
+     */
+    ContinuousQuadratureDataTransfer(const FiniteElement<dim> &projection_fe,
+                                     const Quadrature<dim> &mass_quadrature,
+                                     const Quadrature<dim> &data_quadrature);
+
+    /**
+     * Virtual destructor.
+     */
+    ~ContinuousQuadratureDataTransfer();
+
+    /**
+     * Prepares for coarsening and refinement of a triangulation @p tria .
+     * @p data_storage represents the cell data which should be transfered and
+     * it should be initialized for each locally owned active cell.
+     *
+     * @note Although CellDataStorage class allows storing on different cells
+     * different objects derived from the base class, here we assume that
+     * @p data_storage contains objects of the same type, more specifically
+     * they pack/unpack the same data.
+     */
+    void prepare_for_coarsening_and_refinement (parallel::distributed::Triangulation<dim> &tria,
+                                                CellDataStorage<CellIteratorType,DataType> *data_storage);
```

Comment:
can make it a reference. `tria` can't be a const reference because `register_data_attach()` and `notify_ready_to_unpack()` are not const.